### PR TITLE
feat: Add collection system

### LINKED ISSUE #1318: Multi-dimensional and hierarchical list
## Summary

Update Lists to support hierarchy and multiple dimensions.

## Motivation

Lists support a single dimension list and navigating and selecting within a list. A hierarchical dropdown menu will need an added notion of hierarchy and components like a color picker will need the concept of more than a single dimension.

### PR DESCRIPTION
**Pull request by NicholasBoll**
## Summary

Fixes #1463 
Fixes #1318 

This PR adds a collection system to Canvas Kit! A collection system adds models and elemProps hooks to help create components like tables, menus, selects, etc. The collection system supports selection, navigation, and registration.

category

### Release Note
We've added a collection system to help our users create custom components like 

### BREAKING CHANGES
- Upgrade Typescript to 4.1 and drop support for Typescript <4
- `Popup.Body` is an overflow container. If your popup buttons are inside the `Popup.Body` element (our examples did this), you will need to move those buttons outside the `Popup.Body` element, otherwise the focus rings may be cut off. See migration guide for more information.
- Changed the signature of model callbacks and guards. This is handled by our codemod
  - Before:
    - `shouldUpdate({data, state})`
    - `onUpdate({data, prevState})`
  - After:
    - `shouldUpdate(data, state)`
    - `onUpdate(data, prevState)`

---

## Checklist

- [x] tests are changed or added
- [ ] code has been documented
- [ ] design approved final implementation
- [ ] a11y approved final implementation
- [x] code adheres to the [API & Pattern guidelines](https://workday.github.io/canvas-kit/?path=/story/welcome-dev-docs-api-pattern-guidelines--page)

### CODE REVIEW

File: modules/preview-react/breadcrumbs/lib/Breadcrumbs/Dropdown/Button.tsx
Comment (willklein) [inline modules/preview-react/breadcrumbs/lib/Breadcrumbs/Dropdown/Button.tsx]: Do we need to lock in some Prettier settings?

File: modules/docs/mdx/changelog.stories.mdx
Comment (NicholasBoll) [inline modules/docs/mdx/changelog.stories.mdx:6]: The changelog wasn't working before... This is annoying. Maybe a prettier update can fix this

File: modules/codemod/lib/v7/index.ts
Comment (NicholasBoll) [inline modules/codemod/lib/v7/index.ts:25]: I have no idea why none of these transforms were hooked up yet

Comment (NicholasBoll) [reply modules/codemod/lib/v7/index.ts:25]: https://www.youtube.com/watch?v=0GypdsJulKE

File: modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx
Comment (willklein) [inline modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx:542]: If you're reviewing this, I recommend viewing the Markdown file here: https://github.com/NicholasBoll/canvas-kit/blob/feat/add-grid-support-temp/modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx#model-changes

File: modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx
Comment (willklein) [inline modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx:547]: When you say "intellisense isn't engaged immediately" are we describing that IntelliSense slower with an object parameter?

I will personally complain that with the object, it was one extra level of destructuring to get what I wanted. I agree with making prevState the second parameter too. This makes sense.

File: modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx
Comment (willklein) [inline modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx:581]: Suggestions are not an option for some reason...
`Event maps were a stopgap and were manual and suffered from user-error in defining them.`
could be:
`Event maps were a stopgap, were manual, and suffered from user-error in defining them.`

Comment (willklein) [reply modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx:581]: I had some thoughts about the language choices (`were manual`, `suffered from user-error`) but I think it's fine to let your personal tone shine through!

File: modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx
Comment (willklein) [inline modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx:582]: `events had guards and callbacks`
could be:
`events had guards or callbacks`

File: modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx
Comment (willklein) [inline modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx:585]: > You can still only pass a single argument to events. If you need additional
information, use an object.

I am actually a little confused reading this. I wonder if some of this should move past the next example so the code changes are presented before reading this. This doesn't refer to code in the previous example, does it?

File: modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx
Comment (willklein) [inline modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx:589]: small thing, I don't think `use-cases` needs a hyphen

File: modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx
Comment (NicholasBoll) [reply modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx:547]: Yes. It is an additional level. The parameter given to callbacks and guards was called `event`, and the properties were `data`, and `state` for guards  and `prevState` for callbacks. This often lead to odd paths for events like `show` and `hide` where the parameter for the DOM event was also called `event`: `event.data.event`.

File: modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx
Comment (NicholasBoll) [reply modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx:581]: "prone to errors" is the same idea without sounding like we're pointing fingers

File: modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx
Comment (NicholasBoll) [reply modules/docs/mdx/7.0-MIGRATION-GUIDE.mdx:585]: All events can only take a single parameter because of the signature of guards and callbacks:

```ts
// event signature
type updateValue = (value: string) => void

// guard signature - note the first parameter always gets named `data`
type shouldUpdateValue = (data: string, state: State) => boolean

// callback signature
type onUpdateValue = (data: string, prevState: State) => void
```

The first parameter of the event signature becomes the first parameter of a guard and callback. If we allowed more than a single argument, we'd run into issues:

```ts
// event signature
type updateValue = (value: string, delay = false) => void

// guard - only takes first parameter
type shouldUpdateValue = (data: string, state: State) => boolean

// callback signature - only takes first parameter
type onUpdateValue = (data: string, prevState: State) => void
```

The previous example would have to look like this:
```ts
// event signature
type updateValue = (event: { value: string, delay: boolean }) => void

// guard - only takes first parameter
type shouldUpdateValue = (data: { value: string, delay: boolean }, state: State }) => boolean

// callback signature - only takes first parameter
type onUpdateValue = (data: { value: string, delay: boolean }, prevState: State) => void
```

File: modules/codemod/lib/v7/index.ts
Comment (willklein) [reply modules/codemod/lib/v7/index.ts:25]: I have a pretty good idea why. I'll think about how we can prevent it from slipping again.

File: modules/codemod/lib/v7/updateModelSignatures.ts
Comment (mannycarrera4) [inline modules/codemod/lib/v7/updateModelSignatures.ts:111]: don't forget to add this to the `index` file :)

Comment (NicholasBoll) [reply modules/codemod/lib/v7/updateModelSignatures.ts:111]: Done

File: modules/codemod/lib/v7/index.ts
Comment (NicholasBoll) [reply modules/codemod/lib/v7/index.ts]: One way is to update all codemods to import from this index file instead of their individual transform file. Then the spec won't work until everything is hooked up correctly.

I think we did it this way in v5 because we had cascading transforms that we wanted to isolate, but that might be a mistake. It makes it too easy to have all tests pass, but none of the codemods actually work.

File: modules/react/collection/stories/Collection.stories.mdx
Comment (mannycarrera4) [inline modules/react/collection/stories/Collection.stories.mdx:114]: did you still want this commented out?

File: modules/react/collection/stories/examples/MultiSelection.tsx
Comment (mannycarrera4) [inline modules/react/collection/stories/examples/MultiSelection.tsx:33]: the order of these matters right? is it `useListItemRegister` -> `useListItemRovingFocus` -> `useListItemSelect` where you need to register your items first and your select is last?

File: modules/react/menu/lib/Menu.tsx
Comment (mannycarrera4) [inline modules/react/menu/lib/Menu.tsx:36]: do we need to give it a default element here?

File: modules/react/menu/lib/MenuCard.tsx
Comment (mannycarrera4) [inline modules/react/menu/lib/MenuCard.tsx:52]: I'm a little confused, can you explain why we need elemPropsHook here? It looks like it's returning an empty object?

**Review by mannycarrera4 — APPROVED** [review]
Overall this is huge and super excited to get comfortable with using this, especially with color picker.

A couple notes/ questions?
- when you use `createContainer` do you need to give it a `elemPropsHook` i think I'm a little confused by this. I think it just provides the underlying element the necessary attributes for accessibility is that correct?

- Do we have anywhere written when to use `createComponent` vs `createContainer` vs `createHook` having some sort of guidance would be great, could be something to add after this gets merged

- Do you have to use `createSubComponent` if the parent was created with `createContainer` ? Is this to avoid adding the `model` prop to each sub component?

Just a few questions but awesome work on this!!

File: modules/react/collection/stories/Collection.stories.mdx
Comment (NicholasBoll) [reply modules/react/collection/stories/Collection.stories.mdx:114]: I didn't revisit this after introducing a `ListBox` which takes care of most of this. Now I'm a bit torn on how to document the collection system...

Most cases are covered by `ListBox`, but not all. Originally I was showcasing how all the low-level models and hooks could be used together to create higher-level components. Now I've introduced a few abstractions that cover the common cases I'm outlining and `ListBox` should be used when possible. So... Do I stick with low-level or use the highest abstraction possible?

File: modules/react/collection/stories/examples/DynamicItems.tsx
Comment (NicholasBoll) [reply modules/react/collection/stories/examples/DynamicItems.tsx:17]: We don't need to in this example. `ListBox.Item` already does what we need.

File: modules/react/menu/lib/Menu.tsx
Comment (NicholasBoll) [reply modules/react/menu/lib/Menu.tsx:36]: `Menu` does not represent an element. All containers are responsible for creating a model context, but not all also represent an element. Tabs is similar - according to the specification, there's no wrapper container element needed for tabs to work. This means the container would not represent a semantic element. We've opted to decrease the amount of elements needed for components.

**Comment by NicholasBoll** [comment]
> * when you use `createContainer` do you need to give it a `elemPropsHook` i think I'm a little confused by this. I think it just provides the underlying element the necessary attributes for accessibility is that correct?

`elemPropsHook` only makes sense if your container component represents an element. If there's no element, there's also no elemPropsHook.

> Do we have anywhere written when to use createComponent vs createContainer vs createHook having some sort of guidance would be great, could be something to add after this gets merged

Not yet

> Do you have to use createSubComponent if the parent was created with createContainer ? Is this to avoid adding the model prop to each sub component?

Not necessarily. I'll do a writeup about the transition.

File: modules/react/menu/lib/MenuCard.tsx
Comment (NicholasBoll) [reply modules/react/menu/lib/MenuCard.tsx:52]: It is a placeholder in case `useMenuCard` needs to do something more in the future.

File: modules/react/collection/stories/examples/MultiSelection.tsx
Comment (NicholasBoll) [reply modules/react/collection/stories/examples/MultiSelection.tsx]: This one is weird and I tried to change it when I created it.

It runs hooks from right to left, but merges props from left to right. So the rightmost runs first, while the leftmost props win.

For example, the lodash `compose` works the same way - `compose` runs functions from right to left.

File: modules/codemod/lib/v7/recategorizeIconButtons.ts
Comment (NicholasBoll) [inline modules/codemod/lib/v7/recategorizeIconButtons.ts:124]: Before this change, the following would happen:

```diff
-const MyComponent = createComponent(Button)({
+const MyComponent = createComponent(TerinaryButton)({
```

File: modules/codemod/lib/v7/renameLayoutImports.ts
Comment (NicholasBoll) [inline modules/codemod/lib/v7/renameLayoutImports.ts:82]: The codemod assumed an import would always be found, but it wouldn't if the file under the codemod didn't import from `Layout` which caused the codemod to crash.

File: modules/react/collection/lib/react-virtual.ts
Comment (NicholasBoll) [inline modules/react/collection/lib/react-virtual.ts:2]: Can be removed when https://github.com/TanStack/react-virtual/pull/277 is merged

File: modules/react/collection/lib/useBaseListModel.tsx
Comment (NicholasBoll) [inline modules/react/collection/lib/useBaseListModel.tsx:34]: I'm still working on this, but we don't need to hold up the PR

File: modules/codemod/lib/v7/recategorizeIconButtons.ts
Comment (mannycarrera4) [reply modules/codemod/lib/v7/recategorizeIconButtons.ts:124]: ah so you're adding more specificity?

File: modules/codemod/lib/v7/renameLayoutImports.ts
Comment (mannycarrera4) [reply modules/codemod/lib/v7/renameLayoutImports.ts:82]: ahhhh I also have this PR update this codemod but we can merge this in first! https://github.com/Workday/canvas-kit/pull/1567

File: modules/codemod/lib/v7/recategorizeIconButtons.ts
Comment (NicholasBoll) [reply modules/codemod/lib/v7/recategorizeIconButtons.ts:123]: Yes. It is making sure the function used is named `styled` and the argument is named `IconButton`. Perhaps it should use the importmap instead in case they rename `IconButton`

**Pull request merged**
